### ISSUE TITLE
Add test case for string `eq?` behavior

### ISSUE THREAD

**Issue by eholk** [google/schism#97]
Strings should have the following equality behavior:

```
> (eq? (list->string '(#\a #\b)) (list->string '(#\a #\b)))
#f
> (string=? (list->string '(#\a #\b)) (list->string '(#\a #\b)))
#t
```

Because strings are represented as JavaScript objects, we might be accidentally getting value equality instead of reference equality for `eq?` on strings. We should add a test case to make sure we have the right behavior, and fix it if we don't.

**Comment by wingo** [comment]
Do we actually want this behavior?  Mutable strings are one of the largest design mistakes in Scheme IMO, and unless you need this for some reason, better to avoid.  As you like, of course, I just raise the question :)

**Comment by eholk** [comment]
Ah, fair question. As a default answer, I'd say we should err on the side of complying with a Scheme standard. So far I've been roughly following R6RS, but maybe it's worth following R7RS instead. Or, given how long R7RS Large seems to be taking, maybe we should feel free to break from the standard and just be our own dialect. Regardless, does R6RS require the behavior that two strings created by `list->string` are not `eq?`?

What are your reasons against mutable strings?

**Comment by wingo** [comment]
From a standard POV: all of the scheme standards require that two strings created by make-string, list->string, etc are different in the terms of `eq?`; some functions are defined to create new objects, like `cons`, and `list->string` is one of them.

Scheme as a language standard process has gone as far as it can in a unified way, IMO; the r7/r6 split reflects that pre-existing division.  Each of us is left to pick and choose as we may :)  If I were making a new Scheme implementation, I would make it with immutable pairs, immutable strings, and delimited continuations instead of call/cc.  Mutability is generally not great for reasoning about programs, and confining it to vectors and records is good enough, IMO.  (Like what do you do if a program mutates a list over which you are mapping?)

In the particular case of strings, mutability imposes a lot of implementation restrictions: using the wealth of implementation strategies like cons strings, different encodings, external strings, etc becomes much more expensive.  SRFI-13 defines "string-upcase!", for example, which is unable to upcase "straße" because it would need more characters.  Mutable strings are out of step with language implementation more broadly, and especially so with regards to JavaScript.  I wouldn't go there but that's just me :-)

**Comment by eholk** [comment]
Thanks for the clarifications!

You make a compelling case for being willing to break from the standards as it makes sense. I haven't closely followed the standards process, but I gather that R6RS was contentious, and R7RS-large seems unlikely to finish.

There may not be much reason to do this, but I don't think making `list->string` create a new object necessitates also supporting mutable strings. For example, I might want some kind of unique tag that I can test with `eq?`, but not want to use `gensym` for some reason.

I have "`list->string` makes a new string" pretty firmly embedded in my mental model for Scheme, so my vote would still be to keep that.

I haven't really had any use for mutable strings yet in Schism, or other Scheme implementations that I can think of, so I think it makes sense to hold off on adding them until there's a clear use, if ever. If Schism gets some kind of custom data type support in the future, we might be fine to just add a mutable string library that defines its own parallel string type.

**Comment by eholk** [comment]
Fixed in #98.
